README.md
LGeneral-Deluxe
===============

Open-source reimplementation of Panzer General strategy game.


ChangeLog
- fixed compiler issue in TESTFILE mode (2015/04/03 Michael Tautschnig)

1.2.6:
- fixed all compiler warnings (2015/03/13 MS)
- show message about new unit at beginning of turn (2015/03/13 MS)
- splash/info screen can be closed by window icon (2015/03/13 MS)
- restricted deploy mask: surrounding of own units no longer added to
  prevent moving whole front line one step forward (2015/03/15 MS)
- removed fallback mapnames file from lgc-pg (2015/03/08 Markus Koschany)
- updated icon handling (2015/03/08 Markus Koschany)
- updated desktop file with category and keywords (2015/03/08 Markus Koschany)
- deploy center and damaged info is properly saved/loaded now (2015/03/08 MS)
- added proper engine status for active carpet bombing (2015/03/08 MS)

1.2.5:
- converter reads in core unit flag and core unit limit (2015/03/01 MS)
- fixed bug in carpet bombing: ground units could still supply (2015/02/22 MS)
- carpet bombing only possible by level bombers (2015/02/22 MS)
- added goto unit dialog for easier navigation(2015/02/22 Tez Bartlomiej)
- bombers can temporarily destroy cities/airfields (2015/01/25 Tez Bartlomiej)
- units can be named in scenario files (2015/01/25 Tez Bartlomiej)

1.2.4:
- fixed bad test function declarations (2014/10/18 Michael Tautschnig)
- improved man pages (2012/11/4 Markus Koschany)
- fix broken pg road tile by mirroring counterpart (2012/11/01 MS)
- removed non-free panzer general faq (2012/10/29 MS)
- replaced/modified non-free graphics/sounds (2012/10/29 Markus Koschany)
- applied hardened compile patch (2012/10/11 Markus Koschany)

1.2.3:
- do not allow AI to place units on border of map (2012/04/20 TB)
- assign nations to late-war units: FPO is Poland, FFR is France 
  (2012/04/20 TB)
- fixed bug in final view on scenario: properly disable purchase option
  as no player is selected in this state (2012/04/17 TB)
- added Finnland ids to 'get nation from name' algorithm in lgc-pg:
  it is not used by PG but some additional campaigns (2012/01/27 MS)
- added lgc-pg options to change player names and nations (2012/01/27 MS)
- fixed bug in tactical icon conversion (2012/01/27 MS)
- added some hints about building on Mac OS X (2011/08/30 ACV)
- generate scenario file names from its title for custom campaigns 
  (2011/08/17 MS)

1.2.2:
- applied a number of Gentoo build patches (2011/08/16 MS)
- fixed bug in move range calculation occurring when both, unit and 
  transporter, did not use fuel (2011/08/16 MS)
- fixed buffer overflow in unit name generation (2011/08/16 MS)
- improved conversion of campaigns in lgc-pg (2011/08/09 MS)

1.2.1:
- video mode selection dialog with all available dimensions (2011/06/13 MS)

1.2:
- allow disable/enable of purchase and deploy turn in GUI (10/09/29 MS)
- disable purchase when loading old saved games (10/09/29 MS)
- tweaked initiative bonus for russian tanks (10/09/28 MS)
- use installation path as default for lgc-pg (10/09/28 MS)
- use secure snprintf instead of sprintf in lgc-pg (10/09/28 MS)
- try to open PG files with lowercase and uppercase name (10/09/28 MS)
- added splash screen for lgc-pg (10/09/27 MS)
- added config option for AI debug messages (10/09/27 MS)
- added unit purchase phase for AI (10/09/27 MS)
- gain prestige for capturing flags and damaging units (10/09/26 MS)
- added purchase dialogue (10/09 MS)
- lgc-pg builds prestige data based on original PG data (10/09 MS)
- fixed wrong deploy center block update on loading scenario (10/09/22 MS)
- fixed compilation on 64 bit Linux systems (10/09/20 Pavol Rusnak)
- fixed buffer overflows during campaign loading (10/09/20 Frank Schaefer)

1.2beta-14:
- fixed configure warning (09/10/26 MS)
- added XDG support (09/10/26 MS)
- fixed some errors in READMEs (09/01/09 BH)

1.2beta-13:
- fixed bad color key in font bitmaps (07/09/26 MS)
- added original pg font for campaign briefings (07/09/24 Pamashoid)
- ignore return value in po/lgeneral Makefile (07/09/24 MS)
- bugfix: using transporter crashed if not enough fuel (07/09/24 Triple
Entendre)

1.2beta-12:
- fixed bug in gettext that caused 'make uninstall' to fail (06/05/29 LS)
- fixed a bug that caused end_turn to lock if NUMLOCK was active (06/05/27 MS)
- make it compile under mingw. Thanks to Victor Sergienko. (06/04/10 LS)

1.2beta-11:
- no initial deploy allowed for allied player in MarketGarden (06/01/14 MS)
- fixed translation-encoding by invoking msgconv as necessary (05/12/30 LS)
- updated German translation (05/12/29 LS)
- added option --campaign-start (05/12/29 LS)
- campaign can now be loaded from command-line (05/12/29 LS)
- added full descriptions for PG-campaign (05/12/28 LS)
- brought danger zones in sync with new fuel-usage rules (05/12/28 LS)
- adapted ltrextract to cope with new campaign entries and references
  (05/12/27 LS)
- start campaign with first scenario, not with scenario called "first"
  (05/12/27 LS)
- long-format parser is now more accurate, and stricter (05/12/26 LS)
- added support for title-entry per campaign scenario (05/12/24 LS)
- added support for references in campaign description (05/12/24 LS)
- fixed massive (>30MB per scenario) memory leaks and some smaller ones
  (05/12/21 LS)
- better (and faster) line wrapping algorithm (05/12/20 LS)
- new briefing font (05/12/18 LS)
- allow debriefings in campaigns (05/12/18 LS)
- provide path selection in campaigns (05/12/18 LS)
- explicitly create POFILES and GMOFILES if not defined by gettext.m4
  (05/12/12 LS)

1.2beta-10:
- let some AI units (in town, artillery, close to artillery etc) hold position
  (05/12/09 MS)
- parachutists over an empty allied airfield debark like a normal unit, thus 
  may move on the ground in the same turn (05/12/09 MS)
- on Linux lgeneral is now relocateable (i. e. installation directory
  is not hardcoded anymore) (05/12/07 LS)
- bugfix: don't crash on a savegame called lg_save_11 (05/12/05 LS)
- made savegames interchangeable between big-/little-endian systems
  (05/12/05 LS)

src/map.h
/***************************************************************************
                          map.h  -  description
                             -------------------
    begin                : Sat Jan 20 2001
    copyright            : (C) 2001 by Michael Speck
    email                : kulkanie@gmx.net
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifndef __MAP_H
#define __MAP_H

#include "terrain.h"

enum { FAIR = 0, CLOUDS, RAIN, SNOW };

/*
====================================================================
Map tile
====================================================================
*/
typedef struct {
    char *name;             /* name of this map tile */
    Terrain_Type *terrain;  /* terrain properties */
    int terrain_id;         /* id of terrain properties */
    int image_offset;       /* image offset in prop->image */
    int strat_image_offset; /* offset in the list of strategic tiny terrain images */
    Nation *nation;         /* nation that owns this flag (NULL == no nation) */
    Player *player;         /* dito */
    int obj;                /* military objective ? */
    int deploy_center;      /* deploy allowed? */
    int damaged;  			/* if carped bombed >0, counts down to 0 again */
    Unit *g_unit;           /* ground/naval unit pointer */
    Unit *a_unit;           /* air unit pointer */
} Map_Tile;

/*
====================================================================
To determine various things of the map (deploy, spot, blocked ...)
a map mask is used and these are the flags for it.
====================================================================
*/
enum {
    F_FOG =             ( 1L << 1 ),
    F_SPOT =            ( 1L << 2 ),
    F_IN_RANGE =        ( 1L << 3 ),
    F_MOUNT =           ( 1L << 4 ),
    F_SEA_EMBARK =      ( 1L << 5 ),
    F_AUX =             ( 1L << 6 ),
    F_INFL =            ( 1L << 7 ),
    F_INFL_AIR =        ( 1L << 8 ),
    F_VIS_INFL =        ( 1L << 9 ),
    F_VIS_INFL_AIR =    ( 1L << 10 ),
    F_BLOCKED =         ( 1L << 11 ),
    F_BACKUP =          ( 1L << 12 ),
    F_MERGE_UNIT =      ( 1L << 13),
    F_INVERSE_FOG =     ( 1L << 14 ), /* inversion of F_FOG */
    F_DEPLOY =          ( 1L << 15 ),
    F_CTRL_GRND =       ( 1L << 17 ),
    F_CTRL_AIR =        ( 1L << 18 ),
    F_CTRL_SEA =        ( 1L << 19 ),
    F_MOVE_COST =       ( 1L << 20 ),
    F_DANGER =          ( 1L << 21 ),
    F_SPLIT_UNIT =      ( 1L << 22 ),
    F_DISTANCE =        ( 1L << 23 )
};

/*
====================================================================
Map mask tile.
====================================================================
*/
typedef struct {
    int fog; /* if true the engine covers this tile with fog. if ENGINE_MODIFY_FOG is set
                this fog may change depending on the action (range of unit, merge partners
                etc */
    int spot; /* true if any of your units observes this map tile; you can only attack units
    on a map tile that you spot */
    /* used for a selected unit */
    int in_range; /* this is used for pathfinding; it's -1 if tile isn't in range else it's set to the
    remaining moving points of the unit; enemy influence is not included */
    int distance; /* mere distance to current unit; used for danger mask */
    int moveCost; /* total costs to reach this tile */
    int blocked; /* units can move over there tiles with an allied unit but they must not stop there;
    so allow movment to a tile only if in_range and !blocked */
    int mount; /* true if unit must mount to reach this tile */
    int sea_embark; /* sea embark possible? */
    int infl; /* at the beginning of a player's turn this mask is set; each tile close to a
    hostile unit gets infl increased; if influence is 1 moving costs are doubled; if influence is >=2
    this tile is impassible (unit stops at this tile); if a unit can't see a tile with infl >= 2 and
    tries to move there it will stop on this tile; independed from a unit's moving points passing an
    influenced tile costs all mov-points */
    int vis_infl; /* analogue to infl but only spotted units contribute to this mask; used to setup
    in_range mask */
    int air_infl; /* analouge for flying units */
    int vis_air_infl;
    int aux; /* used to setup any of the upper values */
    int backup; /* used to backup spot mask for undo unit move */
    Unit *merge_unit; /* if not NULL this is a pointer to a unit the one who called map_get_merge_units()
                         may merge with. you'll need to remember the other unit as it is not saved here */
    Unit *split_unit; /* target unit may transfer strength to */
    int  split_okay; /* unit may transfer a new subunit to this tile */
    int deploy; /* deploy mask: 1: unit may deploy their, 0 unit may not deploy their; setup by deploy.c */
    int danger; /* 1: mark this tile as being dangerous to enter */
    /* AI masks */
    int ctrl_grnd; /* mask of controlled area for a player. own units give there positive combat
                      value in move+attack range while enemy scores are substracted. the final
                      value for each tile is relative to the highest absolute control value thus
                      it ranges from -1000 to 1000 */
    int ctrl_air;
    int ctrl_sea; /* each operational region has it's own control mask */
} Mask_Tile;


/*
====================================================================
Load map.
====================================================================
*/
int map_load( char *fname );

/*
====================================================================
Delete map.
====================================================================
*/
void map_delete( );

/*
====================================================================
Get tile at x,y
====================================================================
*/
Map_Tile* map_tile( int x, int y );
Mask_Tile* map_mask_tile( int x, int y );

/*
====================================================================
Clear the passed map mask flags.
====================================================================
*/
void map_clear_mask( int flags );

/*
====================================================================
Swap units. Returns the previous unit or 0 if none.
====================================================================
*/
Unit *map_swap_unit( Unit *unit );

/*
====================================================================
Insert, Remove unit pointer from map.
====================================================================
*/
void map_insert_unit( Unit *unit );
void map_remove_unit( Unit *unit );

/*
====================================================================
Get neighbored tiles clockwise with id between 0 and 5.
====================================================================
*/
Map_Tile* map_get_close_hex( int x, int y, int id );

/*
====================================================================
Add/set spotting of a unit to auxiliary mask
====================================================================
*/
void map_add_unit_spot_mask( Unit *unit );
void map_get_unit_spot_mask( Unit *unit );

/*
====================================================================
Set movement range of a unit to in_range/sea_embark/mount.
====================================================================
*/
void map_get_unit_move_mask( Unit *unit );
void map_clear_unit_move_mask();

/*
====================================================================
Recreates the danger mask for 'unit'.
The fog must be set to the movement range of 'unit' for this
function to work properly.
The movement cost of the mask must have been set for 'unit'.
Returns 1 when at least one tile's danger mask was set, otherwise 0.
====================================================================
*/
int map_get_danger_mask( Unit *unit );

/*
====================================================================
Get a list of way points the unit moves along to it's destination.
This includes check for unseen influence by enemy units (e.g.
Surprise Contact).
====================================================================
*/
typedef struct {
    int x, y;
} Way_Point;
Way_Point* map_get_unit_way_points( Unit *unit, int x, int y, int *count, Unit **ambush_unit );

/*
====================================================================
Backup/restore spot mask to/from backup mask. Used for Undo Turn.
====================================================================
*/
void map_backup_spot_mask();
void map_restore_spot_mask();

/*
====================================================================
Get unit's merge partners and set mask 'merge'.
At maximum MAP_MERGE_UNIT_LIMIT units.
All unused entries in partners are set 0.
====================================================================
*/
enum { MAP_MERGE_UNIT_LIMIT = 6 };
void map_get_merge_units( Unit *unit, Unit **partners, int *count );

/*
====================================================================
Check if unit may transfer strength to unit (if not NULL) or create
a stand alone unit (if unit NULL) on the coordinates.
====================================================================
*/
int map_check_unit_split( Unit *unit, int str, int x, int y, Unit *dest );

/*
====================================================================
Get unit's split partners assuming unit wants to give 'str' strength
points and set mask 'split'. At maximum MAP_SPLIT_UNIT_LIMIT units.
All unused entries in partners are set 0. 'str' must be valid amount,
this is not checked here.
====================================================================
*/
enum { MAP_SPLIT_UNIT_LIMIT = 6 };
void map_get_split_units_and_hexes( Unit *unit, int str, Unit **partners, int *count );

/*
====================================================================
Get a list (vis_units) of all visible units by checking spot mask.
====================================================================
*/
void map_get_vis_units( void );

/*
====================================================================
Draw a map tile terrain to surface. (fogged if mask::fog is set)
====================================================================
*/
void map_draw_terrain( SDL_Surface *surf, int map_x, int map_y, int x, int y );
/*
====================================================================
Draw tile units. If mask::fog is set no units are drawn.
If 'ground' is True the ground unit is drawn as primary
and the air unit is drawn small (and vice versa).
If 'select' is set a selection frame is added.
====================================================================
*/
void map_draw_units( SDL_Surface *surf, int map_x, int map_y, int x, int y, int ground, int select );
/*
====================================================================
Draw danger tile. Expects 'surf' to contain a fully drawn tile at
the given position which will be tinted by overlaying the danger
terrain surface.
====================================================================
*/
void map_apply_danger_to_tile( SDL_Surface *surf, int map_x, int map_y, int x, int y );
/*
====================================================================
Draw terrain and units.
====================================================================
*/
void map_draw_tile( SDL_Surface *surf, int map_x, int map_y, int x, int y, int ground, int select );

/*
====================================================================
Set/update spot mask of by engine's current player.
The update adds the tiles seen by unit.
====================================================================
*/
void map_set_spot_mask();
void map_update_spot_mask( Unit *unit, int *enemy_spotted );

/*
====================================================================
Set mask::fog (which is the actual fog of the engine) to either
spot mask, in_range mask (covers sea_embark), merge mask,
deploy mask.
====================================================================
*/
void map_set_fog( int type );

/*
====================================================================
Set the fog to players spot mask by using mask::aux (not mask::spot)
====================================================================
*/
void map_set_fog_by_player( Player *player );

/*
====================================================================
Check if this map tile is visible to the engine (isn't covered
by mask::fog or mask::spot as modification is allowed and it may be
another's player fog (e.g. one human against cpu))
====================================================================
*/
#define MAP_CHECK_VIS( mapx, mapy ) ( ( !modify_fog && !mask[mapx][mapy].fog ) || ( modify_fog && mask[mapx][mapy].spot ) )

/*
====================================================================
Modify the various influence masks.
====================================================================
*/
void map_add_unit_infl( Unit *unit );
void map_remove_unit_infl( Unit *unit );
void map_remove_vis_unit_infl( Unit *unit );
void map_set_infl_mask();
void map_set_vis_infl_mask();

/*
====================================================================
Check if unit may air/sea embark/debark at x,y.
If 'init' != 0, used relaxed rules for deployment
====================================================================
*/
int map_check_unit_embark( Unit *unit, int x, int y, int type, int init );
int map_check_unit_debark( Unit *unit, int x, int y, int type, int init );

/*
====================================================================
Embark/debark unit and return if an enemy was spotted.
If 'enemy_spotted' is 0, don't recalculate spot mask.
If unit's coordinates or x and y are out of bounds, the respective
tile is not manipulated.
====================================================================
*/
void map_embark_unit( Unit *unit, int x, int y, int type, int *enemy_spotted );
void map_debark_unit( Unit *unit, int x, int y, int type, int *enemy_spotted );

/*
====================================================================
Set the deploy mask for this unit. If 'init', use the initial deploy
mask (or a default one). If not, set the valid deploy centers. In a
second run, remove any tile blocked by an own unit if 'unit' is set.
====================================================================
*/
void map_get_deploy_mask( Player *player, Unit *unit, int init );

/*
====================================================================
Mark this field being a deployment-field for the given player.
====================================================================
*/
void map_set_deploy_field( int mx, int my, int player );

/*
====================================================================
Check whether this field is a deployment-field for the given player.
====================================================================
*/
int map_is_deploy_field( int mx, int my, int player );

/*
====================================================================
Check if unit may be deployed to mx, my or return undeployable unit
there. If 'air_mode' is set the air unit is checked first.
'player' is the index of the player.
====================================================================
*/
int map_check_deploy( Unit *unit, int mx, int my, int player, int init, int air_mode );
Unit* map_get_undeploy_unit( int x, int y, int air_region );

/*
====================================================================
Check the supply level of tile (mx, my) in the context of 'unit'.
(hex tiles with SUPPLY_GROUND have 100% supply)
====================================================================
*/
int map_get_unit_supply_level( int mx, int my, Unit *unit );
/*
====================================================================
Check if this map tile is a supply point for the given unit.
====================================================================
*/
int map_is_allied_depot( Map_Tile *tile, Unit *unit );
/*
====================================================================
Checks whether this hex (mx, my) is supplied by a depot in the
context of 'unit'.
====================================================================
*/
int map_supplied_by_depot( int mx, int my, Unit *unit );

/*
====================================================================
Get drop zone for unit (all close hexes that are free).
====================================================================
*/
void map_get_dropzone_mask( Unit *unit );

#endif
